### ISSUE TITLE
tf.where should have optional dtype parameter

### ISSUE THREAD

**Issue by gkennos** [tensorflow/tensorflow#35141]
- TensorFlow version (you are using): 2.0
- Are you willing to contribute it (Yes/No): Yes

```
tf.keras.backend.floatx = tf.float64
f64 = tf.Variable([0,0.2,0.5,0.7,1], dtype=tf.float64)
tf.where(f64 > 0.5, 1., 0. )

<tf.Tensor: id=16, shape=(5,), dtype=float32, numpy=array([0., 0., 0., 1., 1.], dtype=float32)>
```

alternatives for tf.float64 models lead to difficult to read code

```
tf.where(f64 > 0.5, tf.constant(1., dtype=tf.float64), tf.constant(0., dtype=tf.float64) )
tf.cast(tf.where(f64 > 0.5, 1., 0. ), dtype=tf.float64)
```

assuming that tf.where can't respect the float type of the input variable nor the floatx setting, as those change the api, better would be:

```
tf.where(f64 > 0.5, 1., 0. , dtype=tf.float64)
```

**Comment by jsmeredith** [comment]
Hi!  Thanks for the feedback!

First, I believe that in the listed alternatives, only the first of those two is fully correct.  In the second, there could be e.g. loss of precision in the constants if the cast were to be done after the where().

Which comes back to the first case -- as float32 is the default for TensorFlow, the version `where(condition, tf.constant(x.y, dtype=tf.float64), ...)` is the expected way to do this.  Generally the type is set on the input constant.

So I think the first alternative is correct.  It is probably less common of a problem in practice because constant initialization can happen on its own line -- which also gives them a name, and some implicit meaning, often helping readability.

Having a dtype parameter here would simplify some cases a little as suggested -- not just in where(), but also in other ops.  So if we do start adding a dtype parameter, we should be more intentional about adding it to ops outside of the creation methods.  In other words, think before we would consider adding this we would need to look at the broader spectrum of ops and make sure we know the scope and benefit more holistically.
